File: src/main/java/com/wxl/dyttcrawler/config/EsStoreConfiguration.java
package com.wxl.dyttcrawler.config;

import com.google.common.util.concurrent.ThreadFactoryBuilder;
import com.wxl.dyttcrawler.properties.EsStoreProperties;
import lombok.extern.slf4j.Slf4j;
import org.elasticsearch.client.Node;
import org.elasticsearch.client.RestClient;
import org.springframework.boot.autoconfigure.elasticsearch.RestClientBuilderCustomizer;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

/**
 * Create by wuxingle on 2020/5/10
 * es配置
 */
@Slf4j
@Configuration
@EnableConfigurationProperties(EsStoreProperties.class)
public class EsStoreConfiguration {

    private EsStoreProperties esStoreProperties;

    public EsStoreConfiguration(EsStoreProperties esStoreProperties) {
        this.esStoreProperties = esStoreProperties;
    }

    @Bean
    public RestClientBuilderCustomizer clientBuilderCustomizer() {
        int maxSize = esStoreProperties.getPool().getMaxThreads();

        return builder -> builder.setFailureListener(
                new RestClient.FailureListener() {
                    @Override
                    public void onFailure(Node node) {
                        log.error("es store fail,node is:{}", node);
                    }
                })
                .setHttpClientConfigCallback(httpClientBuilder -> {
                    httpClientBuilder.setMaxConnTotal(maxSize)
                            .setThreadFactory(new ThreadFactoryBuilder()
                                    .setDaemon(true)
                                    .setNameFormat("es-pool-%s")
                                    .build()
                            )
                            .setMaxConnPerRoute(maxSize);
                    return httpClientBuilder;
                })
                .setRequestConfigCallback(requestConfigBuilder -> {
                    requestConfigBuilder.setContentCompressionEnabled(true);
                    return requestConfigBuilder;
                });
    }

}

File: src/main/kotlin/com/wxl/dyttcrawler/config/EsStoreConfiguration.kt
package com.wxl.dyttcrawler.config

import com.google.common.util.concurrent.ThreadFactoryBuilder
import com.wxl.dyttcrawler.properties.EsStoreProperties
import org.elasticsearch.client.Node
import org.elasticsearch.client.RestClient
import org.slf4j.LoggerFactory
import org.springframework.boot.autoconfigure.elasticsearch.RestClientBuilderCustomizer
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

/**
 * Create by wuxingle on 2021/10/11
 * es配置
 */
@Configuration
class EsStoreConfiguration(
    private val esStoreProperties: EsStoreProperties
) {

    companion object {
        private val log = LoggerFactory.getLogger(EsStoreConfiguration::class.java)
    }

    @Bean
    fun clientBuilderCustomizer(): RestClientBuilderCustomizer {
        val maxSize = esStoreProperties.pool.maxThreads

        return RestClientBuilderCustomizer {
            it.setFailureListener(object : RestClient.FailureListener() {
                override fun onFailure(node: Node) {
                    log.error("es store fail,node is:{}", node)
                }
            })
                .setHttpClientConfigCallback { builder ->
                    builder.setMaxConnTotal(maxSize)
                        .setThreadFactory(
                            ThreadFactoryBuilder()
                                .setDaemon(true)
                                .setNameFormat("es-pool-%s")
                                .build()
                        )
                        .setMaxConnPerRoute(maxSize)
                }
                .setRequestConfigCallback { builder ->
                    builder.setContentCompressionEnabled(true)
                }
        }
    }

}